### PR TITLE
ENH: Add support for default implementations.

### PR DESCRIPTION
**Pull request by ssanderson**

### CODE REVIEW

File: interface/default.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        from .typed_signature import TypedSignature
+
+        for default_name, default in keysorted(defaults):
+            impl = default.implementation
+
+            if isinstance(impl, staticmethod):
+                # staticmethods can't use attributes of the interface.
+                continue
+
+            self_name = TypedSignature(impl).first_argument_name
```
Comment (llllllllll) [inline interface/default.py]: if this is a property, we need to check `__get__`, `__set__`, and `__delete__`, right?

File: interface/default.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        " part of {iface}."
+    )
+
+    def warn_if_defaults_use_non_interface_members(interface_name,
+                                                   defaults,
+                                                   members):
+        """Warn if an interface default uses non-interface members of self.
+        """
+        for method_name, attrs in non_member_attributes(defaults, members):
+            warnings.warn(_DEFAULT_USES_NON_INTERFACE_MEMBER_TEMPLATE.format(
```
Comment (llllllllll) [inline interface/default.py]: I wonder if we could have `default` take a list of names to ignore in this warning. One use case I imagine is a default implementation for `__init__` which will set some attribute, then another default implementation that uses that.

File: interface/default.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        it's not guaranteed to find **all** attributes used.
+        """
+        used = set()
+        # Find sequences of the form: LOAD_FAST(local_name), LOAD_ATTR(<name>).
+        # This will find all usages of the form ``local_name.<name>``.
+        #
+        # It will **NOT** find usages in which ``local_name`` is aliased to
+        # another name.
+        for first, second in sliding_window(dis.get_instructions(f), 2):
+            if first.opname == 'LOAD_FAST' and first.argval == local_name:
```
Comment (llllllllll) [inline interface/default.py]: what about store_attr?

File: interface/default.py (hunk 2 above)
Comment (ssanderson) [reply interface/default.py]: I think if you're writing that pattern, then I think you just want an ABC, not an interface.  IMO it should always be safe to swap out the default implementation of a single method without having to know or care about the default implementation of any of the other methods.  If that's not the case, then I don't think the defaults are appropriate.

Comment (llllllllll) [reply interface/default.py]: Okay, that is reasonable.

File: interface/default.py (hunk 1 above)
Comment (ssanderson) [reply interface/default.py]: In general, we only look at `fget` on properties right now.  I could see an argument for expanding to look at other fields, but I think that's a separate PR

File: interface/default.py (hunk 3 above)
Comment (ssanderson) [reply interface/default.py]: Updated this to handle store_attr.

**Pull request merged**